### PR TITLE
Add CodeGenerator tests and fix NREs/bugs

### PR DESCRIPTION
**Pull request by hughbe**
Depends on https://github.com/dotnet/corefx/pull/39585

### CODE REVIEW

**Comment by danmosemsft** [comment]
@buyaa-n your area, could you please help get this in

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs [hunk 1]
```diff
@@ -946,7 +1004,15 @@ protected void GenerateStatements(CodeStatementCollection stmts)
...
         protected virtual void OutputAttributeDeclarations(CodeAttributeDeclarationCollection attributes)
         {
-            if (attributes.Count == 0) return;
+            if (attributes == null)
+            {
+                throw new ArgumentNullException(nameof(attributes));
+            }
+            if (attributes.Count == 0)
+            {
+                return;
```
Comment (krwq) [inline src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]: is this branch fixing any specific issue?

**Comment by krwq** [comment]
product changes LGTM (with one comment)

@hughbe does this need rebase on top of that PR? Presumably some changes will disappear

**Comment by buyaa-n** [comment]
Thank you @hughbe there is some test failures `System.CodeDom.Tests  Total: 3174, Errors: 0, Failed: 61, Skipped: 0, Time: 4.` could take a look, excpet test failures looks great 👍.

**Comment by maryamariyan** [comment]
Thanks for your PR @hughbe 
Just checking if this is still under your radar and if you'll want to update it soon?
Assigned to you so it's easier for you to track.

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs [hunk 2]
```diff
@@ -58,7 +58,10 @@ private void GenerateType(CodeTypeDeclaration e)
                 GenerateDirectives(e.StartDirectives);
             }
 
-            GenerateCommentStatements(e.Comments);
+            if (e.Comments.Count > 0)
+            {
+                GenerateCommentStatements(e.Comments);
+            }
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]: Why is this change necessary?

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs [hunk 3]
```diff
@@ -704,7 +719,15 @@ private void GenerateSnippetMembers(CodeTypeDeclaration e)
...
-            GenerateDirectives(e.StartDirectives);
+            if (e == null)
+            {
+                throw new ArgumentNullException(nameof(e));
+            }
+
+            if (e.StartDirectives.Count > 0)
+            {
+                GenerateDirectives(e.StartDirectives);
+            }
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]: Why is this change necessary?

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs [hunk 4]
```diff
@@ -789,23 +812,48 @@ protected virtual void GenerateCompileUnit(CodeCompileUnit e)
...
-            GenerateCommentStatements(e.Comments);
+            if (e == null)
+            {
+                throw new ArgumentNullException(nameof(e));
+            }
+
+            if (e.Comments.Count > 0)
+            {
+                GenerateCommentStatements(e.Comments);
+            }
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]: Ditto.  I'll stop commenting on these; it's not clear to me why we need to add these Count checks, as all of the called methods handle empty collections just fine.

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs [hunk 5]
```diff
@@ -128,7 +131,10 @@ private void GenerateTypeMember(CodeTypeMember member, CodeTypeDeclaration decla
                 GenerateDirectives(member.StartDirectives);
             }
 
-            GenerateCommentStatements(member.Comments);
+            if (member.Comments.Count > 0)
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]: Why is this change necessary?

At this point in the lifecycle of System.CodeDom, we don't need to make changes like this to try to achieve some kind of artificial consistency.  In fact, we're more likely to break things than to fix things.

**Review by stephentoub — CHANGES_REQUESTED** [review]
Thanks, @hughbe.

I'm fine with the ArgumentNullException changes and the associated tests, but I think the changes that guard calls based on whether collections are empty should be reverted.

**Comment by buyaa-n** [comment]
> I'm fine with the ArgumentNullException changes and the associated tests, but I think the changes that guard calls based on whether collections are empty should be reverted.

@hughbe if you have no time to apply the requested changes in near future please let me know, I can help wiith that. Thank you!

**Comment by danmosemsft** [comment]
@buyaa-n I suggest just making the changes so we can get this one merged...

**Comment by buyaa-n** [comment]
> @buyaa-n I suggest just making the changes so we can get this one merged...

Sure

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]: removed

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs (hunk 5 above)
Comment (buyaa-n) [reply src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]: Removed all count check

**Comment by buyaa-n** [comment]
> I'm fine with the ArgumentNullException changes and the associated tests, but I think the changes that guard calls based on whether collections are empty should be reverted

@stephentoub reverted collection count checks, do you have any other comment?

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

**Comment by hughbe** [comment]
So sorry! I keep getting distracted by winforms stuff and university work. Thanks for taking this on.

**Comment by buyaa-n** [comment]
> So sorry! I keep getting distracted by winforms stuff and university work. Thanks for taking this on.

Sure no problem, thank you for your contribution

**Review by krwq — APPROVED** [review]
Product changes and small tests file LGTM, haven't got time to look at the big tests change

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs [hunk 6]
```diff
@@ -396,6 +401,10 @@ void ICodeGenerator.GenerateCodeFromStatement(CodeStatement e, TextWriter w, Cod
 
         public virtual void GenerateCodeFromMember(CodeTypeMember member, TextWriter writer, CodeGeneratorOptions options)
         {
+            if (member == null)
+            {
+                throw new ArgumentNullException(nameof(member));
+            }
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]:
```suggestion
            }

```

File: src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs [hunk 7]
```diff
@@ -801,11 +820,23 @@ protected virtual void GenerateNamespace(CodeNamespace e)
...
+
             foreach (CodeNamespaceImport imp in e.Imports)
             {
-                if (imp.LinePragma != null) GenerateLinePragmaStart(imp.LinePragma);
+                if (imp.LinePragma != null)
+                {
+                    GenerateLinePragmaStart(imp.LinePragma);
+                }
+
                 GenerateNamespaceImport(imp);
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/CodeGenerator.cs]:
```suggestion
                GenerateNamespaceImport(imp);

```

**Review by stephentoub — APPROVED** [review]
